// @pico-mp/web/dist/index.js
import { CloudTransport } from './transport/CloudTransport.js';
import { MockTransport } from './transport/MockTransport.js';
const receiveHandlers = new Map();
function dispatchEnvelope(senderId, typeName, payload) {
    receiveHandlers.get(typeName)?.forEach(h => h(senderId, payload));
}
// ─── PicoMP public API ────────────────────────────────────────────────────────
let transport = null;
function getTransport() {
    if (!transport)
        throw new Error('[PicoMP] Transport not configured. Call useCloud(), useLocal(), or useMock() first.');
    return transport;
}
function wireTransport(t) {
    t.onEnvelope = dispatchEnvelope;
    t.onPeerJoined = (id) => PicoMP.onPeerJoined?.(id);
    t.onPeerLeft = (id) => PicoMP.onPeerLeft?.(id);
    t.onConnectionLost = () => PicoMP.onConnectionLost?.();
    t.onError = (e, m) => PicoMP.onError?.(e, m);
    t.onRoomDataChanged = (key, value, setBy) => PicoMP.onRoomDataChanged?.(key, value, setBy); // [ADR 0005]
    transport = t;
}
export const PicoMP = {
    // ─── Transport configuration ─────────────────────────────────────────────
    /**
     * Connect to PicoMP Cloud (or any remote Game Server).
     * @param apiKey  Your PicoMP API key
     * @param url     Game Server base URL (default: https://gameserver.pico-mp.com)
     */
    useCloud(apiKey, url = 'https://gameserver.pico-mp.com') {
        wireTransport(new CloudTransport(apiKey, url));
    },
    /**
     * Connect to a locally running Game Server (DEBUG_MODE, wsToken check skipped).
     * @param url  e.g. 'http://localhost:7426'
     */
    useLocal(url) {
        // Local server uses the same protocol; pass a dummy apiKey for DEBUG_MODE
        wireTransport(new CloudTransport('debug', url));
    },
    /** Use in-memory mock transport for offline development. */
    useMock() {
        wireTransport(new MockTransport());
    },
    // ─── Session control ──────────────────────────────────────────────────────
    createRoom(options) {
        return getTransport().createRoom(options);
    },
    joinRoom(roomCode) {
        return getTransport().joinRoom(roomCode);
    },
    quickMatch(options) {
        return getTransport().quickMatch(options);
    },
    leaveRoom() {
        return getTransport().leaveRoom();
    },
    // ─── Messaging ────────────────────────────────────────────────────────────
    /**
     * Send a typed message to a specific peer over WebSocket.
     * @param targetPlayerId  Recipient's playerId
     * @param message         Any serialisable object
     * @param typeName        Type identifier string (e.g. 'PlayerMove')
     */
    send(targetPlayerId, message, typeName) {
        getTransport().sendEnvelope(targetPlayerId, typeName, message);
    },
    /**
     * Broadcast a typed message to all peers in the room.
     * @param message   Any serialisable object
     * @param typeName  Type identifier string (e.g. 'PlayerMove')
     */
    sendAll(message, typeName) {
        getTransport().sendEnvelope(undefined, typeName, message);
    },
    // ─── Receive subscription ─────────────────────────────────────────────────
    /**
     * Subscribe to typed messages.
     * @returns Unsubscribe function — call it to remove the handler.
     *
     * @example
     * const unsub = PicoMP.onReceive<PlayerMove>('PlayerMove', (sender, msg) => { ... });
     * // later:
     * unsub();
     */
    onReceive(typeName, handler) {
        if (typeName.startsWith('__')) {
            console.warn(`[PicoMP] onReceive: '${typeName}'은 시스템 예약 타입입니다. onPeerJoined / onPeerLeft / onConnectionLost 이벤트 핸들러를 사용하세요.`);
            return () => { };
        }
        if (!receiveHandlers.has(typeName))
            receiveHandlers.set(typeName, new Set());
        const set = receiveHandlers.get(typeName);
        const h = handler;
        set.add(h);
        return () => set.delete(h);
    },
    // ─── Event handlers (assign directly) ────────────────────────────────────
    onPeerJoined: null,
    onPeerLeft: null,
    onConnectionLost: null,
    onError: null,
    // ─── Room shared data ─────────────────────────────────────────────────────
    /** Set a room-scoped shared value. All players (including setter) receive onRoomDataChanged. */
    setRoomData(key, value) {
        getTransport().setRoomData(key, value);
    },
    /** Get a room-scoped shared value from local cache. */
    getRoomData(key) {
        return getTransport().getRoomData(key);
    },
    /** Get all room-scoped shared values from local cache. */
    getAllRoomData() {
        return getTransport().getAllRoomData();
    },
    onRoomDataChanged: null,
    // ─── Queries ──────────────────────────────────────────────────────────────
    getPeers() {
        return getTransport().getPeers();
    },
    getMyPlayerId() {
        return transport?.getMyPlayerId() ?? null;
    },
};
