// @pico-mp/web/src/index.ts
import type { ITransport } from './transport/ITransport.js';
import { CloudTransport } from './transport/CloudTransport.js';
import { MockTransport } from './transport/MockTransport.js';
import type { CreateRoomOptions, RoomResult } from './types.js';

// ─── Re-export public types ───────────────────────────────────────────────────
export type { CreateRoomOptions, RoomResult } from './types.js';

// ─── Receive handler registry ────────────────────────────────────────────────
type ReceiveHandler = (senderId: string, payload: unknown) => void;

const receiveHandlers = new Map<string, Set<ReceiveHandler>>();

function dispatchEnvelope(senderId: string, typeName: string, payload: unknown): void {
  receiveHandlers.get(typeName)?.forEach(h => h(senderId, payload));
}

// ─── PicoMP public API ────────────────────────────────────────────────────────
let transport: ITransport | null = null;

function getTransport(): ITransport {
  if (!transport) throw new Error('[PicoMP] Transport not configured. Call useCloud(), useLocal(), or useMock() first.');
  return transport;
}

function wireTransport(t: ITransport): void {
  t.onEnvelope = dispatchEnvelope;
  t.onPeerJoined   = (id) => PicoMP.onPeerJoined?.(id);
  t.onPeerLeft     = (id) => PicoMP.onPeerLeft?.(id);
  t.onConnectionLost = () => PicoMP.onConnectionLost?.();
  t.onError        = (e, m) => PicoMP.onError?.(e, m);
  t.onRoomDataChanged = (key, value, setBy) => PicoMP.onRoomDataChanged?.(key, value, setBy); // [ADR 0005]
  transport = t;
}

export const PicoMP = {
  // ─── Transport configuration ─────────────────────────────────────────────

  /**
   * Connect to PicoMP Cloud (or any remote Game Server).
   * @param apiKey  Your PicoMP API key
   * @param url     Game Server base URL (default: https://gameserver.pico-mp.com)
   */
  useCloud(apiKey: string, url = 'https://gameserver.pico-mp.com'): void {
    wireTransport(new CloudTransport(apiKey, url));
  },

  /**
   * Connect to a locally running Game Server (DEBUG_MODE, wsToken check skipped).
   * @param url  e.g. 'http://localhost:7426'
   */
  useLocal(url: string): void {
    // Local server uses the same protocol; pass a dummy apiKey for DEBUG_MODE
    wireTransport(new CloudTransport('debug', url));
  },

  /** Use in-memory mock transport for offline development. */
  useMock(): void {
    wireTransport(new MockTransport());
  },

  // ─── Session control ──────────────────────────────────────────────────────

  createRoom(options?: CreateRoomOptions): Promise<RoomResult> {
    return getTransport().createRoom(options);
  },

  joinRoom(roomCode: string): Promise<void> {
    return getTransport().joinRoom(roomCode);
  },

  quickMatch(options?: import('./types.js').QuickMatchOptions): Promise<void> {
    return getTransport().quickMatch(options);
  },

  leaveRoom(): Promise<void> {
    return getTransport().leaveRoom();
  },

  // ─── Messaging ────────────────────────────────────────────────────────────

  /**
   * Send a typed message to a specific peer over WebSocket.
   * @param targetPlayerId  Recipient's playerId
   * @param message         Any serialisable object
   * @param typeName        Type identifier string (e.g. 'PlayerMove')
   */
  send<T>(targetPlayerId: string, message: T, typeName: string): void {
    getTransport().sendEnvelope(targetPlayerId, typeName, message);
  },

  /**
   * Broadcast a typed message to all peers in the room.
   * @param message   Any serialisable object
   * @param typeName  Type identifier string (e.g. 'PlayerMove')
   */
  sendAll<T>(message: T, typeName: string): void {
    getTransport().sendEnvelope(undefined, typeName, message);
  },

  // ─── Receive subscription ─────────────────────────────────────────────────

  /**
   * Subscribe to typed messages.
   * @returns Unsubscribe function — call it to remove the handler.
   *
   * @example
   * const unsub = PicoMP.onReceive<PlayerMove>('PlayerMove', (sender, msg) => { ... });
   * // later:
   * unsub();
   */
  onReceive<T>(typeName: string, handler: (senderId: string, msg: T) => void): () => void {
    if (typeName.startsWith('__')) {
      console.warn(`[PicoMP] onReceive: '${typeName}'은 시스템 예약 타입입니다. onPeerJoined / onPeerLeft / onConnectionLost 이벤트 핸들러를 사용하세요.`);
      return () => {};
    }
    if (!receiveHandlers.has(typeName)) receiveHandlers.set(typeName, new Set());
    const set = receiveHandlers.get(typeName)!;
    const h = handler as ReceiveHandler;
    set.add(h);
    return () => set.delete(h);
  },

  // ─── Event handlers (assign directly) ────────────────────────────────────

  onPeerJoined:    null as ((peerId: string) => void) | null,
  onPeerLeft:      null as ((peerId: string) => void) | null,
  onConnectionLost: null as (() => void) | null,
  onError:         null as ((error: string, msg: string) => void) | null,

  // ─── Room shared data ─────────────────────────────────────────────────────

  /** Set a room-scoped shared value. All players (including setter) receive onRoomDataChanged. */
  setRoomData(key: string, value: unknown): void {
    getTransport().setRoomData(key, value);
  },

  /** Get a room-scoped shared value from local cache. */
  getRoomData(key: string): unknown | undefined {
    return getTransport().getRoomData(key);
  },

  /** Get all room-scoped shared values from local cache. */
  getAllRoomData(): Record<string, unknown> {
    return getTransport().getAllRoomData();
  },

  onRoomDataChanged: null as ((key: string, value: unknown, setBy: string) => void) | null,

  // ─── Queries ──────────────────────────────────────────────────────────────

  getPeers(): string[] {
    return getTransport().getPeers();
  },

  getMyPlayerId(): string | null {
    return transport?.getMyPlayerId() ?? null;
  },
};
