// @pico-mp/web/src/transport/CloudTransport.ts
import type { ITransport } from './ITransport.js';
import type { CreateRoomOptions, RoomResult, WsServerFrame, WsSendFrame, WsRoomDataSetFrame } from '../types.js';

function generateId(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID();
  }
  // Fallback for environments without crypto.randomUUID
  return Math.random().toString(36).slice(2) + Date.now().toString(36);
}

/**
 * CloudTransport — connects to a real Game Server over HTTP + WebSocket.
 * Also used for LocalTransport (same protocol, different URL).
 *
 * Protocol reference: ADR 0002, Game_Server/src/http/routes.ts, ws/server.ts
 */
export class CloudTransport implements ITransport {
  private apiKey: string;
  private baseUrl: string;   // http(s)://host
  private wsBaseUrl: string; // ws(s)://host

  private playerId: string | null = null;
  private roomCode: string | null = null;
  private wsToken: string | null = null;
  private ws: WebSocket | null = null;
  private peers: Set<string> = new Set();
  private seq = 0;
  private roomData: Map<string, unknown> = new Map(); // [ADR 0005]

  onPeerJoined: ((peerId: string) => void) | null = null;
  onPeerLeft:   ((peerId: string) => void) | null = null;
  onPeerDisconnected: ((peerId: string) => void) | null = null;
  onPeerReconnected: ((peerId: string) => void) | null = null;
  onConnectionLost: (() => void) | null = null;
  onError: ((error: string, msg: string) => void) | null = null;
  onEnvelope: ((senderId: string, typeName: string, payload: unknown) => void) | null = null;
  onRoomDataChanged: ((key: string, value: unknown, setBy: string) => void) | null = null; // [ADR 0005]

  constructor(apiKey: string, httpUrl: string) {
    this.apiKey = apiKey;
    // Normalise trailing slash
    this.baseUrl = httpUrl.replace(/\/$/, '');
    // Derive WS URL from HTTP URL
    this.wsBaseUrl = this.baseUrl.replace(/^http/, 'ws');
  }

  // ─── HTTP helpers ───────────────────────────────────────────────────────────

  private async post<T>(path: string, body: Record<string, unknown>): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        // [ADR 0002] API key via Authorization: Bearer header (server-preferred path)
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(body),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => '');
      this.onError?.('http_error', `${path} → ${res.status}: ${text}`);
      throw new Error(`HTTP ${res.status} on ${path}: ${text}`);
    }

    return res.json() as Promise<T>;
  }

  // ─── Session control ────────────────────────────────────────────────────────

  async createRoom(options?: CreateRoomOptions): Promise<RoomResult> {
    this.playerId = generateId();
    const data = await this.post<{ roomCode: string; wsToken: string }>('/rooms', {
      playerId: this.playerId,
      options: options ?? {},
    });
    this.roomCode = data.roomCode;
    this.wsToken = data.wsToken;
    await this.connectWs();
    return { roomCode: data.roomCode };
  }

  async joinRoom(roomCode: string): Promise<void> {
    this.playerId = generateId();
    const data = await this.post<{ success: boolean; roomCode: string; wsToken?: string }>(
      '/rooms/join',
      { roomCode, playerId: this.playerId },
    );
    if (!data.success) {
      throw new Error(`joinRoom failed: room ${roomCode} not found or full`);
    }
    this.roomCode = data.roomCode;
    this.wsToken = data.wsToken ?? null;
    await this.connectWs();
  }

  async quickMatch(options?: import('../types.js').QuickMatchOptions): Promise<void> {
    this.playerId = generateId();
    const data = await this.post<{ roomCode: string; wsToken: string; created: boolean }>(
      '/quick-match',
      { playerId: this.playerId, options: options ?? {} },
    );
    this.roomCode = data.roomCode;
    this.wsToken = data.wsToken;
    await this.connectWs();
  }

  async leaveRoom(): Promise<void> {
    if (!this.roomCode || !this.playerId) return;

    try {
      await this.post('/rooms/leave', {
        roomCode: this.roomCode,
        playerId: this.playerId,
      });
    } catch {
      // Best-effort; server cleans up on WS close too
    } finally {
      this.closeWs();
      this.reset();
    }
  }

  // ─── Messaging ──────────────────────────────────────────────────────────────

  sendEnvelope(targetId: string | undefined, typeName: string, payload: unknown): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      this.onError?.('send_error', 'WebSocket is not open');
      return;
    }
    if (!this.playerId) return;

    const frame: WsSendFrame = {
      type: 'send',
      envelope: {
        messageId: generateId(),
        senderId: this.playerId,
        targetId,
        seq: ++this.seq,
        timestampMs: Date.now(),
        content: { type: typeName, payload },
      },
    };
    this.ws.send(JSON.stringify(frame));
  }

  // ─── Room shared data ───────────────────────────────────────────────────────

  setRoomData(key: string, value: unknown): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      this.onError?.('send_error', 'WebSocket is not open');
      return;
    }
    const frame: WsRoomDataSetFrame = { type: 'room_data_set', key, value };
    this.ws.send(JSON.stringify(frame));
  }

  getRoomData(key: string): unknown | undefined {
    return this.roomData.get(key);
  }

  getAllRoomData(): Record<string, unknown> {
    return Object.fromEntries(this.roomData);
  }

  // ─── Queries ────────────────────────────────────────────────────────────────

  getPeers(): string[] {
    return Array.from(this.peers);
  }

  getMyPlayerId(): string | null {
    return this.playerId;
  }

  // ─── WebSocket internals ────────────────────────────────────────────────────

  private connectWs(): Promise<void> {
    if (!this.playerId || !this.roomCode) {
      throw new Error('connectWs called without playerId/roomCode');
    }

    const url = new URL(this.wsBaseUrl);
    url.searchParams.set('playerId', this.playerId);
    url.searchParams.set('roomCode', this.roomCode);
    if (this.wsToken) {
      // [ADR 0002, Game_Server ADR 0010] Session token required for production
      url.searchParams.set('wsToken', this.wsToken);
    }

    return new Promise<void>((resolve, reject) => {
      const ws = new WebSocket(url.toString());
      this.ws = ws;

      const onError = (ev: Event) => {
        reject(new Error('WebSocket connection failed'));
        this.onError?.('ws_error', String(ev));
      };

      ws.onopen = () => {
        ws.removeEventListener('error', onError);
      };

      ws.onerror = onError;

      ws.onmessage = (ev) => {
        let frame: WsServerFrame;
        try {
          frame = JSON.parse(ev.data as string) as WsServerFrame;
        } catch {
          return;
        }
        this.handleFrame(frame, resolve);
      };

      ws.onclose = () => {
        if (this.ws === ws) {
          this.ws = null;
          this.onConnectionLost?.();
        }
      };
    });
  }

  private handleFrame(frame: WsServerFrame, attachedResolve?: () => void): void {
    switch (frame.type) {
      case '__attached':
        // WS is ready — resolve the connectWs promise
        attachedResolve?.();
        break;

      case '__peer_join':
        this.peers.add(frame.playerId);
        this.onPeerJoined?.(frame.playerId);
        break;

      case '__peer_leave':
        this.peers.delete(frame.playerId);
        this.onPeerLeft?.(frame.playerId);
        break;

      case '__peer_disconnected':
        this.onPeerDisconnected?.(frame.playerId);
        break;

      case '__peer_reconnected':
        this.onPeerReconnected?.(frame.playerId);
        break;

      case 'envelope': {
        const { type, payload } = frame.envelope.content ?? {};
        if (typeof type === 'string' && !type.startsWith('__')) {
          this.onEnvelope?.(frame.senderId, type, payload);
        }
        break;
      }

      case '__room_data_changed': // [ADR 0005]
        this.roomData.set(frame.key, frame.value);
        this.onRoomDataChanged?.(frame.key, frame.value, frame.setBy);
        break;

      case '__room_data_sync': // [ADR 0005] Initial sync — populate cache, no callback
        this.roomData.clear();
        for (const [k, v] of Object.entries(frame.data)) {
          this.roomData.set(k, v);
        }
        break;
    }
  }

  private closeWs(): void {
    if (this.ws) {
      this.ws.onclose = null; // suppress onConnectionLost for intentional close
      this.ws.close(1000, 'leaveRoom');
      this.ws = null;
    }
  }

  private reset(): void {
    this.playerId = null;
    this.roomCode = null;
    this.wsToken = null;
    this.peers.clear();
    this.seq = 0;
    this.roomData.clear(); // [ADR 0005]
  }
}
