// @pico-mp/web/src/transport/MockTransport.ts
import type { ITransport } from './ITransport.js';
import type { CreateRoomOptions, RoomResult } from '../types.js';

function generateId(): string {
  return Math.random().toString(36).slice(2) + Date.now().toString(36);
}

/**
 * MockTransport — in-memory simulation for offline development.
 * No network calls. Simulates the Game Server behaviour locally.
 */
export class MockTransport implements ITransport {
  private static rooms: Map<string, Set<string>> = new Map();
  private static roomData: Map<string, Map<string, unknown>> = new Map(); // [ADR 0005] roomCode → key-value

  private playerId: string | null = null;
  private roomCode: string | null = null;
  private peers: Set<string> = new Set();

  onPeerJoined: ((peerId: string) => void) | null = null;
  onPeerLeft:   ((peerId: string) => void) | null = null;
  onPeerDisconnected: ((peerId: string) => void) | null = null; // never fires in Mock
  onPeerReconnected: ((peerId: string) => void) | null = null;  // never fires in Mock
  onConnectionLost: (() => void) | null = null;
  onError: ((error: string, msg: string) => void) | null = null;
  onEnvelope: ((senderId: string, typeName: string, payload: unknown) => void) | null = null;
  onRoomDataChanged: ((key: string, value: unknown, setBy: string) => void) | null = null; // [ADR 0005]

  /** All MockTransport instances sharing the same static room registry */
  private static instances: Set<MockTransport> = new Set();

  constructor() {
    MockTransport.instances.add(this);
  }

  async createRoom(_options?: CreateRoomOptions): Promise<RoomResult> {
    this.playerId = generateId();
    const roomCode = generateId().slice(0, 6).toUpperCase();
    MockTransport.rooms.set(roomCode, new Set([this.playerId]));
    MockTransport.roomData.set(roomCode, new Map()); // [ADR 0005]
    this.roomCode = roomCode;
    return { roomCode };
  }

  async joinRoom(roomCode: string): Promise<void> {
    const room = MockTransport.rooms.get(roomCode);
    if (!room) throw new Error(`Mock room ${roomCode} not found`);

    this.playerId = generateId();
    room.add(this.playerId);
    this.roomCode = roomCode;

    // Notify existing peers (other MockTransport instances in the same room)
    for (const inst of MockTransport.instances) {
      if (inst !== this && inst.roomCode === roomCode) {
        inst.peers.add(this.playerId);
        inst.onPeerJoined?.(this.playerId);
        // Also inform the joining player of existing peers
        this.peers.add(inst.playerId!);
        this.onPeerJoined?.(inst.playerId!);
      }
    }
  }

  async quickMatch(_options?: import('../types.js').QuickMatchOptions): Promise<void> {
    // Find a waiting room with < 4 players
    for (const [code, players] of MockTransport.rooms) {
      if (players.size < 4) {
        return this.joinRoom(code);
      }
    }
    // No room available — create one
    await this.createRoom();
  }

  async leaveRoom(): Promise<void> {
    if (!this.roomCode || !this.playerId) return;

    const room = MockTransport.rooms.get(this.roomCode);
    if (room) {
      room.delete(this.playerId);
      // Notify others
      for (const inst of MockTransport.instances) {
        if (inst !== this && inst.roomCode === this.roomCode) {
          inst.peers.delete(this.playerId!);
          inst.onPeerLeft?.(this.playerId!);
        }
      }
      if (room.size === 0) {
        MockTransport.rooms.delete(this.roomCode);
        MockTransport.roomData.delete(this.roomCode); // [ADR 0005]
      }
    }

    this.peers.clear();
    this.playerId = null;
    this.roomCode = null;
    MockTransport.instances.delete(this);
  }

  sendEnvelope(targetId: string | undefined, typeName: string, payload: unknown): void {
    if (!this.playerId || !this.roomCode) {
      this.onError?.('send_error', 'Not in a room');
      return;
    }

    // Deliver to matching MockTransport instances in the same room
    for (const inst of MockTransport.instances) {
      if (inst === this) continue;
      if (inst.roomCode !== this.roomCode) continue;
      if (targetId && inst.playerId !== targetId) continue;

      // Simulate async delivery
      const sender = this.playerId;
      setTimeout(() => {
        inst.onEnvelope?.(sender, typeName, payload);
      }, 0);
    }
  }

  // [ADR 0005] Room shared data
  setRoomData(key: string, value: unknown): void {
    if (!this.playerId || !this.roomCode) {
      this.onError?.('send_error', 'Not in a room');
      return;
    }
    const store = MockTransport.roomData.get(this.roomCode);
    if (store) store.set(key, value);

    const sender = this.playerId;
    const code = this.roomCode;
    // Deliver to all instances in the room (including setter) asynchronously
    for (const inst of MockTransport.instances) {
      if (inst.roomCode !== code) continue;
      setTimeout(() => {
        inst.onRoomDataChanged?.(key, value, sender);
      }, 0);
    }
  }

  getRoomData(key: string): unknown | undefined {
    if (!this.roomCode) return undefined;
    return MockTransport.roomData.get(this.roomCode)?.get(key);
  }

  getAllRoomData(): Record<string, unknown> {
    if (!this.roomCode) return {};
    const store = MockTransport.roomData.get(this.roomCode);
    if (!store) return {};
    return Object.fromEntries(store);
  }

  getPeers(): string[] {
    return Array.from(this.peers);
  }

  getMyPlayerId(): string | null {
    return this.playerId;
  }

  /** Reset all shared mock state (useful in tests) */
  static reset(): void {
    MockTransport.rooms.clear();
    MockTransport.roomData.clear(); // [ADR 0005]
    MockTransport.instances.clear();
  }
}
