// @pico-mp/web/dist/transport/CloudTransport.js
function generateId() {
    if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
        return crypto.randomUUID();
    }
    // Fallback for environments without crypto.randomUUID
    return Math.random().toString(36).slice(2) + Date.now().toString(36);
}
/**
 * CloudTransport — connects to a real Game Server over HTTP + WebSocket.
 * Also used for LocalTransport (same protocol, different URL).
 *
 * Protocol reference: ADR 0002, Game_Server/src/http/routes.ts, ws/server.ts
 */
export class CloudTransport {
    constructor(apiKey, httpUrl) {
        this.playerId = null;
        this.roomCode = null;
        this.wsToken = null;
        this.ws = null;
        this.peers = new Set();
        this.seq = 0;
        this.roomData = new Map(); // [ADR 0005]
        this.onPeerJoined = null;
        this.onPeerLeft = null;
        this.onPeerDisconnected = null;
        this.onPeerReconnected = null;
        this.onConnectionLost = null;
        this.onError = null;
        this.onEnvelope = null;
        this.onRoomDataChanged = null; // [ADR 0005]
        this.apiKey = apiKey;
        // Normalise trailing slash
        this.baseUrl = httpUrl.replace(/\/$/, '');
        // Derive WS URL from HTTP URL
        this.wsBaseUrl = this.baseUrl.replace(/^http/, 'ws');
    }
    // ─── HTTP helpers ───────────────────────────────────────────────────────────
    async post(path, body) {
        const res = await fetch(`${this.baseUrl}${path}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                // [ADR 0002] API key via Authorization: Bearer header (server-preferred path)
                'Authorization': `Bearer ${this.apiKey}`,
            },
            body: JSON.stringify(body),
        });
        if (!res.ok) {
            const text = await res.text().catch(() => '');
            this.onError?.('http_error', `${path} → ${res.status}: ${text}`);
            throw new Error(`HTTP ${res.status} on ${path}: ${text}`);
        }
        return res.json();
    }
    // ─── Session control ────────────────────────────────────────────────────────
    async createRoom(options) {
        this.playerId = generateId();
        const data = await this.post('/rooms', {
            playerId: this.playerId,
            options: options ?? {},
        });
        this.roomCode = data.roomCode;
        this.wsToken = data.wsToken;
        await this.connectWs();
        return { roomCode: data.roomCode };
    }
    async joinRoom(roomCode) {
        this.playerId = generateId();
        const data = await this.post('/rooms/join', { roomCode, playerId: this.playerId });
        if (!data.success) {
            throw new Error(`joinRoom failed: room ${roomCode} not found or full`);
        }
        this.roomCode = data.roomCode;
        this.wsToken = data.wsToken ?? null;
        await this.connectWs();
    }
    async quickMatch(options) {
        this.playerId = generateId();
        const data = await this.post('/quick-match', { playerId: this.playerId, options: options ?? {} });
        this.roomCode = data.roomCode;
        this.wsToken = data.wsToken;
        await this.connectWs();
    }
    async leaveRoom() {
        if (!this.roomCode || !this.playerId)
            return;
        try {
            await this.post('/rooms/leave', {
                roomCode: this.roomCode,
                playerId: this.playerId,
            });
        }
        catch {
            // Best-effort; server cleans up on WS close too
        }
        finally {
            this.closeWs();
            this.reset();
        }
    }
    // ─── Messaging ──────────────────────────────────────────────────────────────
    sendEnvelope(targetId, typeName, payload) {
        if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
            this.onError?.('send_error', 'WebSocket is not open');
            return;
        }
        if (!this.playerId)
            return;
        const frame = {
            type: 'send',
            envelope: {
                messageId: generateId(),
                senderId: this.playerId,
                targetId,
                seq: ++this.seq,
                timestampMs: Date.now(),
                content: { type: typeName, payload },
            },
        };
        this.ws.send(JSON.stringify(frame));
    }
    // ─── Room shared data ───────────────────────────────────────────────────────
    setRoomData(key, value) {
        if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
            this.onError?.('send_error', 'WebSocket is not open');
            return;
        }
        const frame = { type: 'room_data_set', key, value };
        this.ws.send(JSON.stringify(frame));
    }
    getRoomData(key) {
        return this.roomData.get(key);
    }
    getAllRoomData() {
        return Object.fromEntries(this.roomData);
    }
    // ─── Queries ────────────────────────────────────────────────────────────────
    getPeers() {
        return Array.from(this.peers);
    }
    getMyPlayerId() {
        return this.playerId;
    }
    // ─── WebSocket internals ────────────────────────────────────────────────────
    connectWs() {
        if (!this.playerId || !this.roomCode) {
            throw new Error('connectWs called without playerId/roomCode');
        }
        const url = new URL(this.wsBaseUrl);
        url.searchParams.set('playerId', this.playerId);
        url.searchParams.set('roomCode', this.roomCode);
        if (this.wsToken) {
            // [ADR 0002, Game_Server ADR 0010] Session token required for production
            url.searchParams.set('wsToken', this.wsToken);
        }
        return new Promise((resolve, reject) => {
            const ws = new WebSocket(url.toString());
            this.ws = ws;
            const onError = (ev) => {
                reject(new Error('WebSocket connection failed'));
                this.onError?.('ws_error', String(ev));
            };
            ws.onopen = () => {
                ws.removeEventListener('error', onError);
            };
            ws.onerror = onError;
            ws.onmessage = (ev) => {
                let frame;
                try {
                    frame = JSON.parse(ev.data);
                }
                catch {
                    return;
                }
                this.handleFrame(frame, resolve);
            };
            ws.onclose = () => {
                if (this.ws === ws) {
                    this.ws = null;
                    this.onConnectionLost?.();
                }
            };
        });
    }
    handleFrame(frame, attachedResolve) {
        switch (frame.type) {
            case '__attached':
                // WS is ready — resolve the connectWs promise
                attachedResolve?.();
                break;
            case '__peer_join':
                this.peers.add(frame.playerId);
                this.onPeerJoined?.(frame.playerId);
                break;
            case '__peer_leave':
                this.peers.delete(frame.playerId);
                this.onPeerLeft?.(frame.playerId);
                break;
            case '__peer_disconnected':
                this.onPeerDisconnected?.(frame.playerId);
                break;
            case '__peer_reconnected':
                this.onPeerReconnected?.(frame.playerId);
                break;
            case 'envelope': {
                const { type, payload } = frame.envelope.content ?? {};
                if (typeof type === 'string' && !type.startsWith('__')) {
                    this.onEnvelope?.(frame.senderId, type, payload);
                }
                break;
            }
            case '__room_data_changed': // [ADR 0005]
                this.roomData.set(frame.key, frame.value);
                this.onRoomDataChanged?.(frame.key, frame.value, frame.setBy);
                break;
            case '__room_data_sync': // [ADR 0005] Initial sync — populate cache, no callback
                this.roomData.clear();
                for (const [k, v] of Object.entries(frame.data)) {
                    this.roomData.set(k, v);
                }
                break;
        }
    }
    closeWs() {
        if (this.ws) {
            this.ws.onclose = null; // suppress onConnectionLost for intentional close
            this.ws.close(1000, 'leaveRoom');
            this.ws = null;
        }
    }
    reset() {
        this.playerId = null;
        this.roomCode = null;
        this.wsToken = null;
        this.peers.clear();
        this.seq = 0;
        this.roomData.clear(); // [ADR 0005]
    }
}
